improver/bases.py

```diff
@@ -0,0 +1,261 @@
+# POSSIBILITY OF SUCH DAMAGE.
+"""Module containing plugin base classes."""
+from abc import ABC, abstractmethod
+from typing import List
+
+from iris.cube import Cube, CubeList
+
+import improver.utilities.cube_descriptor
+from improver.utilities.cube_checker import spatial_coords_match
+
+
+class BasePlugin(ABC):
+    """An abstract class for IMPROVER plugins.
+    Subclasses must be callable. We preserve the process
+    method by redirecting to __call__.
+    """
+
+    def __call__(self, *args, **kwargs):
+        """Makes subclasses callable to use process
+        Args:
+            *args:
+                Positional arguments.
+            **kwargs:
+                Keyword arguments.
+        Returns:
+            Output of self.process()
+        """
+        return self.process(*args, **kwargs)
+
+    @abstractmethod
+    def process(self, *args, **kwargs):
+        """Abstract class for rest to implement."""
+        pass
+
+
+class InputCubesPlugin(BasePlugin):
+    """
+    An abstract class for IMPROVER plugins that generate a new diagnostic.
+
+    Sub-classes must set:
+        cube_descriptors: Dict. Each key will be used with get_cube
+        and the value is a CubeDescriptor describing the required cube name and units
+        which the discovered cube will be converted to.
+    """
+
+    def __init__(self, model_id_attr: str = None):
+        """
+        Set up class
+
+        Args:
+            model_id_attr:
+                Name of model ID attribute to be copied from source cubes to output cube
+        """
+        self._parse_cube_descriptors()
+
+        self.model_id_attr = model_id_attr
+        self.model_id_value = None
+
+    def _parse_cube_descriptors(self):
+        """Checks the cube_descriptors dict"""
+        if not self.cube_descriptors:
+            raise ValueError("Missing compulsory dictionary 'cube_descriptors'")
+        for k, v in self.cube_descriptors.items():
+            if not isinstance(k, str):
+                raise TypeError(
+                    f"Keys in cube_descriptors must be 'str', not {type(k)} for {k}"
+                )
+            if not isinstance(v, improver.utilities.cube_descriptor.CubeDescriptor):
+                raise TypeError(
+                    f"Values in cube_descriptors must be <CubeDescriptor>, not {type(v)} for {k}"
+                )
+
+    @property
+    @abstractmethod
+    def cube_descriptors(self):
+        """Classes must set this to a dict where each key will be used with get_cube
+        and the value is a CubeDescriptor describing the required cube name and units
+        which the discovered cube will be converted to."""
+        raise NotImplementedError
+
+    def get_cube(self, key: str) -> Cube:
+        """Gets the named cube.
+
+        Args:
+            key:
+                The cube identifier. Must match a key from cube_descriptors
+        """
+        cube = getattr(self, f"_{key}")
```

Comment:
I think storing these cubes in a dictionary like `self._cubes` might be preferable.
This would also negate checking whether it is an instance of a cube or not (there should be an interface for populating `self._cubes` that would prohibit anything but cubes from populating it.  I would then also just let it raise a keyerror or try-except the keyerror to re-raise with a message with more context.

```python
class InputCubesPlugin(BasePlugin):
    def __init__(...):
        self._cubes = {}
        ...

    def get_cube(self, key: str) -> Cube:
        return self._cubes[key]
```